File: benchmarks/drivers/java/src/main/java/com/mongocore/bench/BenchPipeline.java
package com.mongocore.bench;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.mongocore.MongoClient;
import com.mongocore.Ops;
import mongocore.v1.Mongocore;
import org.bson.Document;
import org.bson.types.ObjectId;

import java.io.FileReader;
import java.io.FileWriter;
import java.lang.management.ManagementFactory;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.util.*;

public class BenchPipeline {
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private static final int[] BATCH_SIZES = {100, 1000, 10000};
    private static final int TOTAL_OPS = 10000;

    static class Config {
        String mongodb_uri;
        String mongocore_address;
        String database;
        int min_time_secs;
        int max_iterations;
        int max_time_secs;
        Map<String, Integer> warmup_iterations;
    }

    static class SystemInfo {
        String os;
        String arch;
        int cpus;
        double ram_gb;
        String mongocore_version;
        String driver;

        SystemInfo() {
            this.os = System.getProperty("os.name").toLowerCase();
            this.arch = System.getProperty("os.arch");
            this.cpus = Runtime.getRuntime().availableProcessors();
            long memory = ((com.sun.management.OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean()).getTotalMemorySize();
            this.ram_gb = Math.round(memory / (1024.0 * 1024.0 * 1024.0) * 10) / 10.0;
            this.mongocore_version = "0.6.0";
            this.driver = "mongocore+java";
        }
    }

    static class Percentiles {
        double p10, p25, p50, p75, p90, p95, p99;
    }

    static class BenchResult {
        String benchmark;
        String category;
        String driver;
        int dataset_size_bytes;
        int batch_size;
        int iterations;
        double total_time_secs;
        double ops_per_sec;
        double mb_per_sec;
        Percentiles percentiles;
        String timestamp;
        SystemInfo system;
    }

    interface TaskFn {
        void run(MongoClient client) throws Exception;
    }

    private static double percentile(List<Double> data, int pct) {
        int idx = (int) Math.ceil(data.size() * pct / 100.0) - 1;
        if (idx < 0) idx = 0;
        if (idx >= data.size()) idx = data.size() - 1;
        return data.get(idx);
    }

    private static BenchResult runBenchmark(
            String name,
            String category,
            MongoClient client,
            TaskFn setupFn,
            TaskFn beforeTaskFn,
            TaskFn taskFn,
            TaskFn afterTaskFn,
            TaskFn teardownFn,
            int datasetSizeBytes,
            int batchSize,
            Config config
    ) throws Exception {
        setupFn.run(client);

        int warmup = config.warmup_iterations.get("java");
        for (int i = 0; i < warmup; i++) {
            beforeTaskFn.run(client);
            taskFn.run(client);
            afterTaskFn.run(client);
        }

        List<Double> times = new ArrayList<>();
        double totalTime = 0.0;
        int iteration = 0;

        while (totalTime < config.min_time_secs || iteration < 5) {
            if (iteration >= config.max_iterations || totalTime >= config.max_time_secs) {
                break;
            }

            beforeTaskFn.run(client);
            long start = System.nanoTime();
            taskFn.run(client);
            double elapsed = (System.nanoTime() - start) / 1_000_000_000.0;
            afterTaskFn.run(client);

            times.add(elapsed);
            totalTime += elapsed;
            iteration++;
        }

        teardownFn.run(client);

        Collections.sort(times);
        double median = times.get(times.size() / 2);
        double opsPerSec = batchSize / median;
        double mbPerSec = datasetSizeBytes / median / 1_000_000.0;

        BenchResult result = new BenchResult();
        result.benchmark = name;
        result.category = category;
        result.driver = "mongocore+java";
        result.dataset_size_bytes = datasetSizeBytes;
        result.batch_size = batchSize;
        result.iterations = times.size();
        result.total_time_secs = Math.round(totalTime * 1000) / 1000.0;
        result.ops_per_sec = Math.round(opsPerSec * 10) / 10.0;
        result.mb_per_sec = Math.round(mbPerSec * 1000) / 1000.0;

        Percentiles pct = new Percentiles();
        pct.p10 = Math.round(percentile(times, 10) * 1_000_000) / 1_000_000.0;
        pct.p25 = Math.round(percentile(times, 25) * 1_000_000) / 1_000_000.0;
        pct.p50 = Math.round(median * 1_000_000) / 1_000_000.0;
        pct.p75 = Math.round(percentile(times, 75) * 1_000_000) / 1_000_000.0;
        pct.p90 = Math.round(percentile(times, 90) * 1_000_000) / 1_000_000.0;
        pct.p95 = Math.round(percentile(times, 95) * 1_000_000) / 1_000_000.0;
        pct.p99 = Math.round(percentile(times, 99) * 1_000_000) / 1_000_000.0;
        result.percentiles = pct;

        result.timestamp = Instant.now().toString();
        result.system = new SystemInfo();

        System.out.printf("  %s: %.0f ops/s, %.2f MB/s (%d iterations)%n",
                name, opsPerSec, mbPerSec, times.size());
        return result;
    }

    public static void main(String[] args) throws Exception {
        System.out.println("=== MongoCore+Java Pipeline benchmarks ===");

        boolean quickMode = java.util.Arrays.asList(args).contains("--quick");

        Path configPath = Paths.get("..", "common.json");
        Config config = GSON.fromJson(new FileReader(configPath.toFile()), Config.class);

        if (quickMode) {
            config.warmup_iterations.put("java", 0);
            config.min_time_secs = 0;
            config.max_iterations = 1;
            config.max_time_secs = 5;
        }

        Path dataDir = Paths.get("..", "..", "data");
        String smallDocJson = Files.readString(dataDir.resolve("small_doc.json"));
        String tweetDocJson = Files.readString(dataDir.resolve("tweet.json"));

        Document smallDoc = Document.parse(smallDocJson);
        Document tweetDoc = Document.parse(tweetDocJson);

        int smallSize = smallDocJson.getBytes().length;
        int tweetSize = tweetDocJson.getBytes().length;

        MongoClient client = MongoClient.create(config.mongocore_address);
        List<BenchResult> results = new ArrayList<>();

        for (int bs : BATCH_SIZES) {
            int callsPerIter = TOTAL_OPS / bs;

            // --- pipeline_run_command ---
            final int fbs = bs;
            final int fcalls = callsPerIter;
            results.add(runBenchmark(
                    "pipeline_run_command_" + bs, "pipeline", client,
                    c -> {},
                    c -> {},
                    c -> {
                        for (int call = 0; call < fcalls; call++) {
                            Mongocore.PipelineOperation[] ops = new Mongocore.PipelineOperation[fbs];
                            for (int i = 0; i < fbs; i++) {
                                ops[i] = Ops.runCommand(config.database, new Document("hello", 1), false);
                            }
                            c.pipeline(ops);
                        }
                    },
                    c -> {},
                    c -> {},
                    TOTAL_OPS * 100, TOTAL_OPS, config
            ));

            // --- pipeline_insert_one_small ---
            String collInsert = "bench_pipeline_insert_java";
            results.add(runBenchmark(
                    "pipeline_insert_one_small_" + bs, "pipeline", client,
                    c -> {},
                    c -> {
                        try { c.runCommand(config.database, new Document("drop", collInsert), false); } catch (Exception ignored) {}
                    },
                    c -> {
                        for (int call = 0; call < fcalls; call++) {
                            Mongocore.PipelineOperation[] ops = new Mongocore.PipelineOperation[fbs];
                            for (int i = 0; i < fbs; i++) {
                                Document doc = new Document(smallDoc);
                                doc.put("_id", new ObjectId().toHexString());
                                ops[i] = Ops.insert(config.database, collInsert, doc);
                            }
                            c.pipeline(ops);
                        }
                    },
                    c -> {},
                    c -> {},
                    TOTAL_OPS * smallSize, TOTAL_OPS, config
            ));

            // --- pipeline_find_one_by_id ---
            String collFind = "bench_pipeline_find_java";
            results.add(runBenchmark(
                    "pipeline_find_one_by_id_" + bs, "pipeline", client,
                    c -> {
                        try { c.runCommand(config.database, new Document("drop", collFind), false); } catch (Exception ignored) {}
                        Document doc = new Document(tweetDoc);
                        doc.put("_id", "bench_find_001");
                        c.getDatabase(config.database).getCollection(collFind).insertOne(doc);
                    },
                    c -> {},
                    c -> {
                        for (int call = 0; call < fcalls; call++) {
                            Mongocore.PipelineOperation[] ops = new Mongocore.PipelineOperation[fbs];
                            for (int i = 0; i < fbs; i++) {
                                ops[i] = Ops.findOne(config.database, collFind, new Document("_id", "bench_find_001"));
                            }
                            c.pipeline(ops);
                        }
                    },
                    c -> {},
                    c -> {
                        c.runCommand(config.database, new Document("drop", collFind), false);
                    },
                    TOTAL_OPS * tweetSize, TOTAL_OPS, config
            ));
        }

        client.close();

        Path resultsDir = Paths.get("..", "..", "results");
        Files.createDirectories(resultsDir);
        Path outputPath = resultsDir.resolve("java_pipeline.json");
        try (FileWriter writer = new FileWriter(outputPath.toFile())) {
            GSON.toJson(results, writer);
        }
        System.out.printf("%nResults saved to %s%n", outputPath);
    }
}

File: benchmarks/drivers/go/bench_pipeline.go
// Benchmark MongoCore Go pipeline batching at different batch sizes.
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"time"

	"github.com/rozza/mongocore/clients/go/mongocore"
	"github.com/rozza/mongocore/clients/go/mongocore/ops"
	pb "github.com/rozza/mongocore/clients/go/proto"
	"go.mongodb.org/mongo-driver/v2/bson"
)

type Config struct {
	MongoDBURI    string         `json:"mongodb_uri"`
	MongoCoreAddr string         `json:"mongocore_address"`
	Database      string         `json:"database"`
	MinTimeSecs   int            `json:"min_time_secs"`
	MaxIterations int            `json:"max_iterations"`
	MaxTimeSecs   int            `json:"max_time_secs"`
	WarmupIters   map[string]int `json:"warmup_iterations"`
}

type SystemInfo struct {
	OS           string `json:"os"`
	Arch         string `json:"arch"`
	CPUs         int    `json:"cpus"`
	MongoCoreVer string `json:"mongocore_version"`
	Driver       string `json:"driver"`
}

type Percentiles struct {
	P10 float64 `json:"p10"`
	P25 float64 `json:"p25"`
	P50 float64 `json:"p50"`
	P75 float64 `json:"p75"`
	P90 float64 `json:"p90"`
	P95 float64 `json:"p95"`
	P99 float64 `json:"p99"`
}

type BenchResult struct {
	Benchmark     string      `json:"benchmark"`
	Category      string      `json:"category"`
	Driver        string      `json:"driver"`
	DatasetBytes  int         `json:"dataset_size_bytes"`
	BatchSize     int         `json:"batch_size"`
	Iterations    int         `json:"iterations"`
	TotalTimeSecs float64     `json:"total_time_secs"`
	OpsPerSec     float64     `json:"ops_per_sec"`
	MBPerSec      float64     `json:"mb_per_sec"`
	Percentiles   Percentiles `json:"percentiles"`
	Timestamp     string      `json:"timestamp"`
	System        SystemInfo  `json:"system"`
}

var batchSizes = []int{100, 1000, 10000}

const totalOps = 10000

func getSystemInfo() SystemInfo {
	return SystemInfo{
		OS:           runtime.GOOS,
		Arch:         runtime.GOARCH,
		CPUs:         runtime.NumCPU(),
		MongoCoreVer: "0.6.0",
		Driver:       "mongocore+go",
	}
}

func percentile(data []float64, pct int) float64 {
	idx := int(math.Ceil(float64(len(data))*float64(pct)/100.0)) - 1
	if idx < 0 {
		idx = 0
	}
	if idx >= len(data) {
		idx = len(data) - 1
	}
	return data[idx]
}

func runBenchmark(
	name, category string,
	setupFn func() error,
	beforeTaskFn func() error,
	taskFn func() error,
	afterTaskFn func() error,
	teardownFn func() error,
	datasetBytes, batchSize int,
	config Config,
) BenchResult {
	if err := setupFn(); err != nil {
		panic(err)
	}

	warmup := config.WarmupIters["go"]
	for i := 0; i < warmup; i++ {
		_ = beforeTaskFn()
		_ = taskFn()
		_ = afterTaskFn()
	}

	times := []float64{}
	totalTime := 0.0
	iteration := 0

	for totalTime < float64(config.MinTimeSecs) || iteration < 5 {
		if iteration >= config.MaxIterations || totalTime >= float64(config.MaxTimeSecs) {
			break
		}

		_ = beforeTaskFn()
		start := time.Now()
		if err := taskFn(); err != nil {
			panic(err)
		}
		elapsed := time.Since(start).Seconds()
		_ = afterTaskFn()

		times = append(times, elapsed)
		totalTime += elapsed
		iteration++
	}

	_ = teardownFn()

	sort.Float64s(times)
	median := times[len(times)/2]
	opsPerSec := float64(batchSize) / median
	mbPerSec := float64(datasetBytes) / median / 1_000_000

	result := BenchResult{
		Benchmark:     name,
		Category:      category,
		Driver:        "mongocore+go",
		DatasetBytes:  datasetBytes,
		BatchSize:     batchSize,
		Iterations:    len(times),
		TotalTimeSecs: math.Round(totalTime*1000) / 1000,
		OpsPerSec:     math.Round(opsPerSec*10) / 10,
		MBPerSec:      math.Round(mbPerSec*1000) / 1000,
		Percentiles: Percentiles{
			P10: math.Round(percentile(times, 10)*1_000_000) / 1_000_000,
			P25: math.Round(percentile(times, 25)*1_000_000) / 1_000_000,
			P50: math.Round(median*1_000_000) / 1_000_000,
			P75: math.Round(percentile(times, 75)*1_000_000) / 1_000_000,
			P90: math.Round(percentile(times, 90)*1_000_000) / 1_000_000,
			P95: math.Round(percentile(times, 95)*1_000_000) / 1_000_000,
			P99: math.Round(percentile(times, 99)*1_000_000) / 1_000_000,
		},
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		System:    getSystemInfo(),
	}

	fmt.Printf("  %s: %.0f ops/s, %.2f MB/s (%d iterations)\n",
		name, opsPerSec, mbPerSec, len(times))
	return result
}

func isQuickMode() bool {
	for _, arg := range os.Args[1:] {
		if arg == "--quick" {
			return true
		}
	}
	return false
}

func main() {
	fmt.Println("=== MongoCore+Go Pipeline benchmarks ===")

	configPath := filepath.Join("..", "common.json")
	configData, err := os.ReadFile(configPath)
	if err != nil {
		panic(err)
	}
	var config Config
	json.Unmarshal(configData, &config)

	if isQuickMode() {
		config.WarmupIters["go"] = 0
		config.MinTimeSecs = 0
		config.MaxIterations = 1
		config.MaxTimeSecs = 5
	}

	dataDir := filepath.Join("..", "..", "data")
	smallDocData, _ := os.ReadFile(filepath.Join(dataDir, "small_doc.json"))
	tweetDocData, _ := os.ReadFile(filepath.Join(dataDir, "tweet.json"))

	var smallDoc map[string]interface{}
	var tweetDoc map[string]interface{}
	json.Unmarshal(smallDocData, &smallDoc)
	json.Unmarshal(tweetDocData, &tweetDoc)

	smallSize := len(smallDocData)
	tweetSize := len(tweetDocData)

	client := mongocore.MongoClientTCP(config.MongoCoreAddr)
	ctx := context.Background()
	if err := client.Connect(ctx); err != nil {
		panic(err)
	}
	defer client.Close()

	results := []BenchResult{}

	for _, bs := range batchSizes {
		callsPerIter := totalOps / bs

		// --- pipeline_run_command ---
		results = append(results, runBenchmark(
			fmt.Sprintf("pipeline_run_command_%d", bs), "pipeline",
			func() error { return nil },
			func() error { return nil },
			func() error {
				for c := 0; c < callsPerIter; c++ {
					pipelineOps := make([]*pb.PipelineOperation, bs)
					for i := range pipelineOps {
						op, err := ops.RunCommand(config.Database, bson.D{{"hello", 1}}, false)
						if err != nil {
							return err
						}
						pipelineOps[i] = op
					}
					_, err := client.Pipeline(ctx, pipelineOps...)
					if err != nil {
						return err
					}
				}
				return nil
			},
			func() error { return nil },
			func() error { return nil },
			totalOps*100, totalOps, config,
		))

		// --- pipeline_insert_one_small ---
		collInsert := "bench_pipeline_insert_go"
		results = append(results, runBenchmark(
			fmt.Sprintf("pipeline_insert_one_small_%d", bs), "pipeline",
			func() error { return nil },
			func() error {
				client.RunCommand(ctx, config.Database, bson.D{{"drop", collInsert}}, false)
				return nil
			},
			func() error {
				for c := 0; c < callsPerIter; c++ {
					pipelineOps := make([]*pb.PipelineOperation, bs)
					for i := range pipelineOps {
						doc := bson.D{{"_id", bson.NewObjectID().Hex()}}
						for k, v := range smallDoc {
							doc = append(doc, bson.E{Key: k, Value: v})
						}
						op, err := ops.Insert(config.Database, collInsert, doc)
						if err != nil {
							return err
						}
						pipelineOps[i] = op
					}
					_, err := client.Pipeline(ctx, pipelineOps...)
					if err != nil {
						return err
					}
				}
				return nil
			},
			func() error { return nil },
			func() error { return nil },
			totalOps*smallSize, totalOps, config,
		))

		// --- pipeline_find_one_by_id ---
		collFind := "bench_pipeline_find_go"
		results = append(results, runBenchmark(
			fmt.Sprintf("pipeline_find_one_by_id_%d", bs), "pipeline",
			func() error {
				client.RunCommand(ctx, config.Database, bson.D{{"drop", collFind}}, false)
				doc := bson.D{{"_id", "bench_find_001"}}
				for k, v := range tweetDoc {
					doc = append(doc, bson.E{Key: k, Value: v})
				}
				coll := client.Database(config.Database).Collection(collFind)
				_, err := coll.InsertOne(ctx, doc)
				return err
			},
			func() error { return nil },
			func() error {
				for c := 0; c < callsPerIter; c++ {
					pipelineOps := make([]*pb.PipelineOperation, bs)
					for i := range pipelineOps {
						op, err := ops.FindOne(config.Database, collFind, bson.D{{"_id", "bench_find_001"}})
						if err != nil {
							return err
						}
						pipelineOps[i] = op
					}
					_, err := client.Pipeline(ctx, pipelineOps...)
					if err != nil {
						return err
					}
				}
				return nil
			},
			func() error { return nil },
			func() error {
				client.RunCommand(ctx, config.Database, bson.D{{"drop", collFind}}, false)
				return nil
			},
			totalOps*tweetSize, totalOps, config,
		))
	}

	resultsDir := filepath.Join("..", "..", "results")
	os.MkdirAll(resultsDir, 0755)
	outputPath := filepath.Join(resultsDir, "go_pipeline.json")
	data, _ := json.MarshalIndent(results, "", "  ")
	os.WriteFile(outputPath, data, 0644)
	fmt.Printf("\nResults saved to %s\n", outputPath)
}
